File: lib/cryptdatum/bench/cmd/cryptdatum-bench.go
// Copyright 2022 The howijd.network Authors
// Licensed under the Apache License, Version 2.0.
// See the LICENSE file.

package main

import (
	"errors"
	"io"
	"log"
	"os"

	"howijd.network/lib/cryptdatum"
)

func main() {
	switch os.Args[1] {
	case "verify":
		cmdVerify(os.Args[2])
	default:
		log.Fatal("invalid command")
	}
}

func cmdVerify(file string) {
	if len(os.Args) < 2 {
		log.Fatal("error: no subcommand provided.")
	}
	ctd, err := os.Open(file)
	if err != nil {
		log.Fatal(err)
	}
	defer ctd.Close()
	headb := make([]byte, cryptdatum.HeaderSize)

	if _, err := ctd.Read(headb); err != nil && !errors.Is(err, io.EOF) {
		log.Fatal(err)
	}
	if !cryptdatum.VerifyHeader(headb) {
		os.Exit(1)
	}
	os.Exit(0)
}

File: lib/cryptdatum/bench/cmd/cryptdatum-bench.rs
use std::env;
use std::error::Error;
use std::fs::File;
use std::io::Read;
use std::process::exit;

use cryptdatum::verify_header;

fn main() -> Result<(), Box<dyn Error>> {
  let args: Vec<String> = env::args().collect();

  if args.len() < 2 {
      println!("error: no subcommand provided.");
      exit(1);
  }

  let command = &args[1];
  let file = &args[2];

  match command.as_str() {
      "verify" => cmd_verify(file)?,
      _ => {
          println!("invalid command");
          exit(1);
      }
  }

  Ok(())
}

fn cmd_verify(file: &str) -> Result<(), Box<dyn Error>> {
  let mut ctd = File::open(file)?;
  let mut headb = [0; cryptdatum::HEADER_SIZE];

  ctd.read_exact(&mut headb)?;

  if !verify_header(&headb) {
      exit(1);
  }

  Ok(())
}
